File: scripts/pocket2/duml.py
# SPDX-License-Identifier: GPL-3.0-or-later
"""
DUML v1 — DJI's device command protocol — encode and decode.

Layout (from the publicly documented Wireshark dissector in dji-firmware-tools):

  0      SOF 0x55
  1..2   length (10 bits, low) | version (6 bits, high), little-endian
  3      CRC8 over bytes 0..2
  4      sender:   type (bits 0-4) | index (bits 5-7)      -- per comm_mkdupc.py
  5      receiver: type (bits 0-4) | index (bits 5-7)
  6..7   sequence, little-endian
  8      cmd type: bit7 = 1 for a response, bits 5-6 ack wanted, bits 0-2 encryption
  9      cmdset
  10     cmdid
  11..   payload
  last 2 CRC16 over everything before it, little-endian

Device types: 1 camera, 2 app, 4 gimbal, 6 remote, 10 PC. Command sets: 0 general,
2 camera, 4 gimbal. The CRC seeds are DJI's: CRC8 reflected poly 0x8C seed 0x77, CRC16 reflected poly
0x1021 (table 0000 1189 2312 …) seed 0x3692 — both checked against dji-firmware-tools.
"""
import struct

SOF = 0x55
DEV_CAMERA, DEV_APP, DEV_GIMBAL, DEV_RC, DEV_PC = 1, 2, 4, 6, 10
CMDSET = {0: "general", 1: "special", 2: "camera", 3: "flightctl", 4: "gimbal",
          6: "rc", 7: "wifi", 13: "battery"}

def _table8(poly=0x8C):
    # reflected form of 0x31 — the reference table starts 00 5e bc e2 61 3f dd 83
    t = []
    for i in range(256):
        c = i
        for _ in range(8):
            c = (c >> 1) ^ poly if c & 1 else c >> 1
        t.append(c)
    return t

def _table16(poly=0x1021):
    # reflected table for the KERMIT family
    t = []
    for i in range(256):
        c = i
        for _ in range(8):
            c = (c >> 1) ^ 0x8408 if c & 1 else c >> 1
        t.append(c)
    return t

_T8, _T16 = _table8(), _table16()

def crc8(data: bytes, seed=0x77) -> int:
    c = seed
    for b in data:
        c = _T8[(c ^ b) & 0xFF]
    return c

def crc16(data: bytes, seed=0x3692) -> int:
    c = seed
    for b in data:
        c = (c >> 8) ^ _T16[(c ^ b) & 0xFF]
    return c & 0xFFFF

def dev(type_, index=0) -> int:
    return (type_ & 0x1F) | ((index & 0x7) << 5)

def encode(cmdset: int, cmdid: int, payload: bytes = b"", *, seq: int = 0,
           sender=DEV_APP, receiver=DEV_CAMERA, sender_idx=1, receiver_idx=0,
           response=False, ack=1, encrypt=0, version=1) -> bytes:
    """ack=1 asks the device to acknowledge; ack=0 is fire-and-forget."""
    length = 13 + len(payload)
    head = bytes([SOF]) + struct.pack("<H", (length & 0x3FF) | ((version & 0x3F) << 10))
    head += bytes([crc8(head)])
    body = bytes([dev(sender, sender_idx), dev(receiver, receiver_idx)])
    body += struct.pack("<H", seq & 0xFFFF)
    body += bytes([((1 if response else 0) << 7) | ((ack & 3) << 5) | (encrypt & 7),
                   cmdset & 0xFF, cmdid & 0xFF]) + payload
    frame = head + body
    return frame + struct.pack("<H", crc16(frame))

class Frame:
    __slots__ = ("raw", "length", "version", "sender", "sender_idx", "receiver",
                 "receiver_idx", "seq", "response", "encrypt", "ack", "cmdset", "cmdid",
                 "payload", "crc_ok")
    def __repr__(self):
        who = lambda t, i: f"{ {1:'cam',2:'app',4:'gim',6:'rc',10:'pc'}.get(t, t)}{i}"
        kind = "RSP" if self.response else "REQ"
        cs = CMDSET.get(self.cmdset, str(self.cmdset))
        ok = "" if self.crc_ok else " CRC-BAD"
        pl = self.payload.hex(" ") if len(self.payload) <= 40 else self.payload[:40].hex(" ") + " …"
        return (f"{kind} {who(self.sender, self.sender_idx)}→{who(self.receiver, self.receiver_idx)} "
                f"seq={self.seq} ack={self.ack} {cs}/0x{self.cmdid:02x} len={len(self.payload)}{ok} [{pl}]")

def decode(buf: bytes) -> Frame:
    f = Frame(); f.raw = buf
    lv = struct.unpack_from("<H", buf, 1)[0]
    f.length, f.version = lv & 0x3FF, lv >> 10
    f.sender, f.sender_idx = buf[4] & 0x1F, buf[4] >> 5
    f.receiver, f.receiver_idx = buf[5] & 0x1F, buf[5] >> 5
    f.seq = struct.unpack_from("<H", buf, 6)[0]
    ct = buf[8]
    f.response, f.ack, f.encrypt = bool(ct & 0x80), (ct >> 5) & 3, ct & 7
    f.cmdset, f.cmdid = buf[9], buf[10]
    f.payload = bytes(buf[11:f.length - 2])
    f.crc_ok = crc8(buf[:3]) == buf[3] and crc16(buf[:f.length - 2]) == struct.unpack_from("<H", buf, f.length - 2)[0]
    return f

class Splitter:
    """Feed arbitrary byte chunks from a bulk pipe; yields complete DUML frames.
    Bytes that are not a frame (some links wrap DUML in their own envelope) are
    surfaced as ('junk', bytes) so nothing is silently dropped."""
    def __init__(self):
        self.buf = bytearray()
    def feed(self, data: bytes):
        self.buf += data
        while True:
            i = self.buf.find(SOF)
            if i < 0:
                if self.buf: junk, self.buf = bytes(self.buf), bytearray(); yield ("junk", junk)
                return
            if i:
                junk = bytes(self.buf[:i]); del self.buf[:i]; yield ("junk", junk)
            if len(self.buf) < 4:
                return
            if crc8(bytes(self.buf[:3])) != self.buf[3]:
                junk = bytes(self.buf[:1]); del self.buf[:1]; yield ("junk", junk); continue
            length = struct.unpack_from("<H", self.buf, 1)[0] & 0x3FF
            if length < 13:
                junk = bytes(self.buf[:1]); del self.buf[:1]; yield ("junk", junk); continue
            if len(self.buf) < length:
                return
            frame = bytes(self.buf[:length]); del self.buf[:length]
            yield ("frame", decode(frame))

if __name__ == "__main__":
    # self-test: a frame must decode to what was encoded, with both CRCs valid
    p = encode(0, 1, b"", seq=7)
    f = decode(p)
    assert f.crc_ok and f.cmdset == 0 and f.cmdid == 1 and f.seq == 7, f
    print("ok", p.hex(" "), f)

File: packages/yonder-core/src/video/accessory/duml.ts
// SPDX-License-Identifier: GPL-3.0-or-later

/**
 * DUML v1, DJI's device command protocol.
 *
 * The wire layout and CRC parameters are ported from
 * `scripts/pocket2/duml.py`. That reference records their provenance in the
 * public dji-firmware-tools Wireshark dissector and `comm_mkdupc.py`:
 *
 *   0      SOF 0x55
 *   1..2   length (10 bits) | version (6 bits), little endian
 *   3      CRC8 over bytes 0..2
 *   4..5   sender and receiver type (low 5 bits), index (high 3 bits)
 *   6..7   sequence, little endian
 *   8      response bit 7, acknowledgement bits 5..6, encryption bits 0..2
 *   9..10  command set and command id
 *   11..   payload
 *   last 2 CRC16 over everything before it, little endian
 */

export const DUML_SOF = 0x55;
export const MIN_DUML_FRAME = 13;
export const MAX_DUML_FRAME = 0x3ff;
export const MAX_DUML_PAYLOAD = MAX_DUML_FRAME - MIN_DUML_FRAME;

export const DEV_CAMERA = 1;
export const DEV_APP = 2;
export const DEV_GIMBAL = 4;
export const DEV_RC = 6;
export const DEV_PC = 10;

const CRC8_SEED = 0x77;
const CRC16_SEED = 0x3692;

export function crc8(data: Uint8Array, seed = CRC8_SEED): number {
  let crc = seed & 0xff;
  for (const byte of data) {
    crc ^= byte;
    for (let bit = 0; bit < 8; bit += 1) {
      crc = (crc & 1) !== 0 ? (crc >>> 1) ^ 0x8c : crc >>> 1;
    }
  }
  return crc & 0xff;
}

export function crc16(data: Uint8Array, seed = CRC16_SEED): number {
  let crc = seed & 0xffff;
  for (const byte of data) {
    crc ^= byte;
    for (let bit = 0; bit < 8; bit += 1) {
      crc = (crc & 1) !== 0 ? (crc >>> 1) ^ 0x8408 : crc >>> 1;
    }
  }
  return crc & 0xffff;
}

export interface DumlCommand {
  readonly commandSet: number;
  readonly commandId: number;
  readonly payload?: Uint8Array;
  readonly sequence?: number;
  readonly sender?: number;
  readonly senderIndex?: number;
  readonly receiver?: number;
  readonly receiverIndex?: number;
  readonly response?: boolean;
  readonly ack?: number;
  readonly encryption?: number;
  readonly version?: number;
}

export interface DumlFrame {
  readonly raw: Uint8Array;
  readonly length: number;
  readonly version: number;
  readonly sender: number;
  readonly senderIndex: number;
  readonly receiver: number;
  readonly receiverIndex: number;
  readonly sequence: number;
  readonly response: boolean;
  readonly ack: number;
  readonly encryption: number;
  readonly commandSet: number;
  readonly commandId: number;
  readonly payload: Uint8Array;
}

function integer(name: string, value: number, minimum: number, maximum: number): number {
  if (!Number.isInteger(value) || value < minimum || value > maximum) {
    throw new RangeError(`${name} must be an integer from ${minimum} to ${maximum}`);
  }
  return value;
}

function device(type: number, index: number): number {
  return integer("device type", type, 0, 0x1f)
    | (integer("device index", index, 0, 7) << 5);
}

export function encodeDuml(command: DumlCommand): Uint8Array {
  const payload = command.payload?.slice() ?? new Uint8Array(0);
  if (payload.length > MAX_DUML_PAYLOAD) {
    throw new RangeError(`payload must be at most ${MAX_DUML_PAYLOAD} bytes`);
  }

  const length = MIN_DUML_FRAME + payload.length;
  const version = integer("version", command.version ?? 1, 0, 0x3f);
  const sequence = integer("sequence", command.sequence ?? 0, 0, 0xffff);
  const ack = integer("ack", command.ack ?? 1, 0, 3);
  const encryption = integer("encryption", command.encryption ?? 0, 0, 7);
  const commandSet = integer("commandSet", command.commandSet, 0, 0xff);
  const commandId = integer("commandId", command.commandId, 0, 0xff);

  const frame = new Uint8Array(length);
  const view = new DataView(frame.buffer);
  frame[0] = DUML_SOF;
  view.setUint16(1, length | (version << 10), true);
  frame[3] = crc8(frame.subarray(0, 3));
  frame[4] = device(command.sender ?? DEV_APP, command.senderIndex ?? 1);
  frame[5] = device(command.receiver ?? DEV_CAMERA, command.receiverIndex ?? 0);
  view.setUint16(6, sequence, true);
  frame[8] = (command.response === true ? 0x80 : 0) | (ack << 5) | encryption;
  frame[9] = commandSet;
  frame[10] = commandId;
  frame.set(payload, 11);
  view.setUint16(length - 2, crc16(frame.subarray(0, length - 2)), true);
  return frame;
}

/** Strictly decode exactly one complete, valid DUML frame. */
export function decodeDuml(input: Uint8Array): DumlFrame | null {
  if (input.length < MIN_DUML_FRAME || input.length > MAX_DUML_FRAME) return null;
  if (input[0] !== DUML_SOF || crc8(input.subarray(0, 3)) !== input[3]) return null;

  const source = input.slice();
  const view = new DataView(source.buffer, source.byteOffset, source.byteLength);
  const lengthVersion = view.getUint16(1, true);
  const length = lengthVersion & 0x3ff;
  if (length !== source.length || length < MIN_DUML_FRAME) return null;
  const sentCrc = view.getUint16(length - 2, true);
  if (crc16(source.subarray(0, length - 2)) !== sentCrc) return null;

  const commandType = source[8]!;
  return {
    raw: source,
    length,
    version: lengthVersion >>> 10,
    sender: source[4]! & 0x1f,
    senderIndex: source[4]! >>> 5,
    receiver: source[5]! & 0x1f,
    receiverIndex: source[5]! >>> 5,
    sequence: view.getUint16(6, true),
    response: (commandType & 0x80) !== 0,
    ack: (commandType >>> 5) & 3,
    encryption: commandType & 7,
    commandSet: source[9]!,
    commandId: source[10]!,
    payload: source.slice(11, length - 2),
  };
}

export type DumlSplitPart =
  | { readonly kind: "frame"; readonly frame: DumlFrame }
  | { readonly kind: "junk"; readonly data: Uint8Array };

function concat(left: Uint8Array, right: Uint8Array): Uint8Array {
  const joined = new Uint8Array(left.length + right.length);
  joined.set(left);
  joined.set(right, left.length);
  return joined;
}

/** Reassembles frames from arbitrary bulk chunks without trusting a bad length. */
export class DumlSplitter {
  private buffer: Uint8Array = new Uint8Array(0);

  get bufferedBytes(): number { return this.buffer.length; }

  reset(): void { this.buffer = new Uint8Array(0); }

  push(chunk: Uint8Array): DumlSplitPart[] {
    this.buffer = concat(this.buffer, chunk);
    const parts: DumlSplitPart[] = [];

    while (this.buffer.length > 0) {
      const start = this.buffer.indexOf(DUML_SOF);
      if (start < 0) {
        parts.push({ kind: "junk", data: this.buffer });
        this.buffer = new Uint8Array(0);
        break;
      }
      if (start > 0) {
        parts.push({ kind: "junk", data: this.buffer.slice(0, start) });
        this.buffer = this.buffer.slice(start);
      }
      if (this.buffer.length < 4) break;

      if (crc8(this.buffer.subarray(0, 3)) !== this.buffer[3]) {
        parts.push({ kind: "junk", data: this.buffer.slice(0, 1) });
        this.buffer = this.buffer.slice(1);
        continue;
      }
      const length = (this.buffer[1]! | (this.buffer[2]! << 8)) & 0x3ff;
      if (length < MIN_DUML_FRAME) {
        parts.push({ kind: "junk", data: this.buffer.slice(0, 1) });
        this.buffer = this.buffer.slice(1);
        continue;
      }
      if (this.buffer.length < length) break;

      const frame = decodeDuml(this.buffer.subarray(0, length));
      if (frame === null) {
        parts.push({ kind: "junk", data: this.buffer.slice(0, 1) });
        this.buffer = this.buffer.slice(1);
        continue;
      }
      parts.push({ kind: "frame", frame });
      this.buffer = this.buffer.slice(length);
    }
    return parts;
  }
}
